src/Block/Service/AbstractMenuBlockService.php

```diff
@@ -0,0 +1,246 @@
+use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormTypeInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Twig\Environment;
+
+abstract class AbstractMenuBlockService extends AbstractBlockService implements EditableBlockService
+{
+    public function __construct(Environment $twig)
+    {
+        parent::__construct($twig);
+    }
+
+    public function execute(BlockContextInterface $blockContext, ?Response $response = null): Response
+    {
+        $template = $blockContext->getTemplate();
+        \assert(null !== $template);
+
+        $responseSettings = [
+            'menu' => $this->getMenu($blockContext),
+            'menu_options' => $this->getMenuOptions($blockContext->getSettings()),
+            'block' => $blockContext->getBlock(),
+            'context' => $blockContext,
+        ];
+
+        // NEXT_MAJOR: remove
+        if ('private' === $blockContext->getSetting('cache_policy')) {
+            return $this->renderPrivateResponse($template, $responseSettings, $response);
+        }
+
+        return $this->renderResponse($template, $responseSettings, $response);
+    }
+
+    public function configureCreateForm(FormMapper $form, BlockInterface $block): void
+    {
+        $this->configureEditForm($form, $block);
+    }
+
+    public function configureEditForm(FormMapper $form, BlockInterface $block): void
+    {
+        $form->add('settings', ImmutableArrayType::class, [
+            'keys' => $this->getFormSettingsKeys(),
+            'translation_domain' => 'SonataBlockBundle',
+        ]);
+    }
+
+    public function validate(ErrorElement $errorElement, BlockInterface $block): void
+    {
+    }
+
+    /**
+     * NEXT_MAJOR: Remove this method.
+     */
+    public function getMetadata(): MetadataInterface
+    {
+        return new Metadata('sonata.block.service.menu', null, null, 'SonataBlockBundle', [
+            'class' => 'fa fa-bars',
+        ]);
+    }
+
+    public function configureSettings(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'title' => '',
+            // NEXT_MAJOR: Remove.
+            'cache_policy' => 'public',
+            'template' => '@SonataBlock/Block/block_core_menu.html.twig',
+            'safe_labels' => false,
+            'current_class' => 'active',
+            'first_class' => false,
+            'last_class' => false,
+            // NEXT_MAJOR: Remove.
+            'current_uri' => null,
+            // NEXT_MAJOR: Remove.
+            'menu_class' => 'list-group',
+            // NEXT_MAJOR: Remove.
+            'children_class' => 'list-group-item',
+            'menu_template' => null,
+        ]);
+
+        // NEXT_MAJOR: Remove setDeprecated.
+        $resolver->setDeprecated(
```

Comment:
This looks wrong. The class is brand new and introduces deprecations?

Can you move the options and deprecations to the using class.

Reply:
Does this make sense to you @VincentLanglet ?